### DISCUSSION TITLE
Trouble with state management, data loading, and auth

### DISCUSSION THREAD

**Discussion by pilcrowonpaper** [nuxt/nuxt#21895] [General]
I maintain an auth library and I've been trying to create a basic example with Nuxt.

My experience so far has been pretty frustrating, which is unfortunate. Nuxt has been pretty different from other meta frameworks I've used (SvelteKit, Astro, Next.js, Remix), and it's possible that my desired approach just doesn't work with Nuxt. 

I think the best way to explain my issue is by showing what I usually do in other frameworks. Here, I create a *protected* page that gets the current user and their posts:

```ts
// runs in the server, before the page is rendered
const load = async () => {
  const user = await validateSession(); // validates session cookie with a db call
  if (!user) return redirect("/login"); // redirect unauthenticated users to login page
  const posts = await getUserPosts(user.userId);
  return {
    user,
    posts
  }
};
```
```svelte
<!-- svelte component -->
<script lang="ts">
  export let data: PageData;
  const { user, posts } = data;
</script>
```

Each page gets its own dedicated function (`load()`) for loading data, which runs before the page renders. The result of the function is accessible in the page component (`+page.svelte`). While I used SvelteKit as an example, other frameworks has similar APIs (Next.js' `getServerSideProps()`, React with server components, Remix' `loader()`, and Astro's frontmatter).

There are few points I'd like to highlight.

1. Avoid duplicate fetch requests: The user and user's data (`posts`) can be fetched in a single API request.
2. Properly typed auth state: `user` is always defined, and typed as such (never `User | null`).
3. Data validity and invalidation: The auth state will be always be updated when the user navigates, and I can always re-run the load function to get the latest data

I've been trying to find a way to replicate this approach in Nuxt, with all 3 benefits mentioned above, but I have yet to find a clean solution.

## Avoid duplicate fetch requests

The solution I went with is to store and share the current user using `useState()`. That skips the fetch request to get the user. Even if the user was authenticated, the next request to fetch the user's posts will return a 401/403 error.

```ts
const useUser = async () => {
     const user = useState<User | null | undefined>("user", () => undefined);
     if (user.value === undefined) {
          const { data, error } = await useFetch("/api/user");
          if (error.value) throw createError("Failed to fetch data");
          user.value = data.value?.user ?? null;
    }
     return user
};

const user = await useUser();
const { data, error } = await useFetch("/api/user");
if (error.value) throw createError("Unauthenticated");
```

But, I don't really like that I can't use an async function to populate the initial value, and more importantly, using shared states brings its own complexity and issues. 

## Properly typed auth state

Middleware seems to be the recommend way to protect pages.

```ts
definePageMeta({
  middleware: ["auth"]
});
```

Even if this page is protected, `useUser()` won't know that and the ref will always be typed `User | null`. This is an issue when using middleware-like API in any frameworks I've used, but it's usually solved by using load functions... which Nuxt doesn't have. I can't narrow types for `ref` values, and even if I use `unref()`, I can't return or throw a redirect inside `setup`.

## Data invalidation

How do I invalidate the user state when the state changes? The obvious answer is to just override the `user` ref:

```ts
const user = useAuth()
const handleLogin = async () => {
  const authenticatedUser = await login(credentials);
  user.value = authenticatedUser;
  await navigateTo("/profile");
}
```

But, that's not how apps really work. As a user, I expect to be redirected first and then the new page reflects my new auth state. It's more obvious for sign outs:

```ts
const user = useAuth()
const handleLogout = async () => {
  await logout();
  user.value = null;
  await navigateTo("/profile");
}
```

The protected page, which displays some user data (like profile picture), goes blank before navigating to the login page. So I guess what I need to do here is flag the data as outdated, and on the next page navigation (to `/login`), refetch the data. 

## Custom `load()` function

One idea I has was to create my load function. In essence, each page will get its own API route. But again, there's no way to redirect unauthenticated users, and returning a redirect response from API routes won't work either (due to how `fetch()` works).

## Temporary solution

Here's my temporary solution. `useUser()` always returns the user, authenticated or unauthenticated. `useAuthenticatedUser()` will always return the authenticated user, but will throw when used in a non-protected page. This fixes the type issues. I also store the validity of the user state, and the user state can be invalidated anytime. I feel this is more complicated than it needs to be, and I really hope there's an easier way to handle this.

```ts
export const invalidateUserState = async () => {
   const userState = useState<"invalid" | "valid">(
    "user_state",
    () => "invalid"
  );
   userState.value = "invalid";
};

export const useUser = async () => {
   const userState = useState<"invalid" | "valid">(
    "user_state",
    () => "invalid"
  );
   const user = useState<User | null>("user", () => null);
   if (userState.value === "invalid") {
      userState.value = "valid";
      const { data, error } = await useFetch("/api/user");
      if (error.value) throw createError("Failed to fetch data");
      user.value = data.value?.user ?? null;
  }
   return computed(() => user.value);  // readonly ref
};

export const useAuthenticatedUser = async () => {
   const user = unref(await useUser());
   if (!user) {
      throw createError(
        "useAuthenticatedUser() can only be used in protected pages"
      );
  }
   const authenticatedUser = useState("authenticated_user", () => user);
   return computed(() => authenticatedUser); // readonly ref
};
```

**Comment by Hebilicious** [comment]
Thanks so much for this write-up, this is very valuable feedback that will help improve the framework.
I think you make some excellent points. 

We should definitely add more fully fledged examples around authentication, integrating with popular libraries such as auth.js, lucia, laravel, but also with saas providers like auth0/clerk/hank0. 

While Nuxt 3 doesn't have an official auth module like Nuxt2 did, there's many modules out there that already help to do these things the Nuxt way, for better discoverability, we could add an authentication section to the docs, and add authentication examples to the main repo.

Regarding your specific points : 

- Loader functions : Many community members like this pattern too, and some modules will provide you with a similar experience : https://github.com/wobsoriano/numix , https://github.com/Hebilicious/form-actions-nuxt . I hope we can integrate this pattern in Nuxt directly 🙏🏽 

- Redirecting within API routes : Right, out of the box, there's no way to have redirects that works both on SSR and CSR from API routes. While this is solvable by things such as `form-actions-nuxt`(https://github.com/Hebilicious/form-actions-nuxt/blob/9d0ce08ef5dbee3d5fde20c0013b18fc37ac1bed/packages/form-actions-nuxt/src/runtime/composables/useFormAction.ts#L88), this should come with the framework.

- Using a composable to handle redirects : This is an interesting approach that has some nice benefits. You could use route middleware to simplify the code. But I agree that it's not obvious and different from other meta frameworks.

- Type issue :  I agree and I think it would be very nice if we could use the `throw`  keyword with a redirect/navigateTo, which would make it easier to create typesafe composables: 

```ts
export const useUser = async () => {
  const { data } = await useFetch("/api/user") //or validate your user
  const user = toValue(data) 
  if(!user) throw navigateTo("/") // this doesn't work but would be very nice for typesafety.
  // you could add loading the posts here
  return readonly(user)
};
```
Not sure if that's a good idea, as it can cause infinite loops if you use this in "/". 

One of the reasons that I personally love Nuxt is that this complexity can be abstracted from end users with the module system. I'll probably release a Lucia module to make it 0 config and easy for anyone to use the library.

Still, all of what you said applies and we should simplify streamline some of these things, to make library authors life easier and end-users who don't want to use modules too.

**Comment by thejspro** [comment]
I'm experiencing similar problems. State-sharing between the client and the server is a fundamental piece of functionality and it worked great in Nuxt 2, with Nuxt 3 it's completely broken - or at least the documentation is not clear enough.

Please do something about it, as I am a huge fan on Vue and Nuxt, but Nuxt 3 and Vue 3 are simply *not production ready*.

**Reply by danielroe** [reply]
Would you open an issue with what isn't working for you, so we can improve the docs and make the DX better?

**Reply by thejspro** [reply]
Thanks for getting back to me so quickly, Daniel.

I think you're doing a fantastic job with Nuxt, and I can only imagine how much work and dedication has gone into it, however the primary selling point of Vue and Nuxt 2 was simplicity, and with Nuxt 3 some key areas are not easy to grasp, especially when switching from Nuxt 2.

There should be a clear and concise explanation of how to achieve things such as 'nuxtServerInit' or SSR and client state-sharing in Nuxt 3. We desperately need an official authentication plugin - the one created by sidebase does not work and is too convoluted. These are just a few things I've experienced when switching from 2 to 3, but I'm sure others have experienced similar problems as well.

There are several issues and discussions asking for clarification on how to do this or how to do that or what is the best practice and the truth is, that with the composition API it's free for all. There are several ways to skin a cat.

I guess what we, as the community are asking for, is some official guidance on what would be considered Nuxt 3 best practices and how to achieve what was possible in Nuxt 2 with Nuxt 3.

The migration guide only scratches the surface and there are plenty of questions unanswered.

I don't want to criticise, I'm merely hoping that all the knowledge in your head can somehow be transferred to the documentation.

Thank you!

**Reply by aaronksaunders** [reply]
@goodpixels - hey i was reading your comment about NuxtAuth from SideBase and how it doesn't work
>We desperately need an official authentication plugin - the one created by sidebase does not work and is too convoluted.

I have been playing around with it since I plan to use it in an upcoming project, What exactly is not working for you? I would like to find an alternative if there is going to be a blocking issue in the future.

thanks
